### PR TITLE
Updates to steampipe service. Closes #523.

### LINKED ISSUE #523: `service` refactoring and performance improvements

### PR DESCRIPTION
**Pull request by binaek89**
### Multiple invocations
Multiple instances of `steampipe query` and `steampipe check` can run, even if the background DB service was started by either - we refer to this type as an **implicit service**.

When an instance of `steampipe query` or `steampipe check` exits, it checks whether the background DB service has active connections (other than itself) and shuts down the service if and only if there are none - exception being when the service was explicitly started by `steampipe service start` (referred to as **explicit service**)

### `service` Conversion
During `steampipe service start` if there is an active **implicit service** which was started by `query` or `check`, it is now converted to an **explicit service** and is not shutdown even if the _invoking_ `steampipe query` or `steampipe check` session exits.

#### Multiple `steampipe query` sessions
It is possible to run multiple `steampipe query` sessions simultaneously without having to run a `steampipe service start` 

#### Multiple `steampipe check` sessions
It is possible to run multiple `steampipe check` sessions simultaneously without having to run a `steampipe service start` 

#### Parallel `steampipe query` and `steampipe check` sessions
Parallel sessions of `steampipe query` and `steampipe check` is now possible.

> In all of the above cases, the last session to quit will shutdown the `service` upon exit.
> *Limitation:* If other client's are connected to the `service`, then shutting down the `service` is skipped.

#### `service start` conversion
When service is running **implicitly**, `service start` converts the service to an **explicit** service, if the values of `--database-listen` and `--database-port` match to the running service.
```
binaeksarkar@Binaeks-MacBook-Pro sample_workspace % steampipe service start --database-port=9194     
Error: service is already running on port 9193 - cannot change port while it's running
binaeksarkar@Binaeks-MacBook-Pro sample_workspace % 
```
```
binaeksarkar@Binaeks-MacBook-Pro sample_workspace % steampipe service start --database-listen=local     
Error: service is already running and listening on network - cannot change listen type while it's running
binaeksarkar@Binaeks-MacBook-Pro sample_workspace % 
```
#### `service start --foreground`
```
binaeksarkar@Binaeks-MacBook-Pro sample_workspace % steampipe service start --foreground 

Steampipe database service is now running:

	Host(s):  localhost, 127.0.0.1
	Port:     9193
	Database: steampipe
	User:     steampipe
	Password: ****-****-****

Connection string:

	postgres://steampipe:****-****-****@localhost:9193/steampipe?sslmode=disable

Managing Steampipe service:

	# Get status of the service
	steampipe service status
	
	# Restart the service
	steampipe service restart

	# Stop the service
	steampipe service stop
	

Hit Ctrl+C to stop the service


```

#### `service status --all`
```
binaeksarkar@Binaeks-MacBook-Pro sample_workspace % steampipe service status --all
+-------+--------------------------------+------+--------+
| PID   | Install Directory              | Port | Listen |
+-------+--------------------------------+------+--------+
| 55576 | /Users/binaeksarkar/.steampipe | 9193 | local  |
+-------+--------------------------------+------+--------+
binaeksarkar@Binaeks-MacBook-Pro sample_workspace % 
```

#### `service stop` fails if clients are connected
`service stop` does not shutdown the service if clients are connected to the service.
```
binaeksarkar@Binaeks-MacBook-Pro ~ % steampipe service stop

Cannot stop service since there are clients connected to the service.

To force stop the service, use steampipe service stop --force

binaeksarkar@Binaeks-MacBook-Pro ~ % 
```

#### `service stop` fails if service was started implicitly
If the service was started by any command other than `steampipe service start` then calling `steampipe service stop` on said service will fail with the message (when started by `query`):
```
binaeksarkar@Binaeks-MacBook-Pro ~ % steampipe service stop

Steampipe service is running exclusively for an active steampipe query session.

To force stop the service, use steampipe service stop --force

binaeksarkar@Binaeks-MacBook-Pro ~ % 
```

### CODE REVIEW

File: cmd/query.go [hunk 1]
```diff
@@ -114,6 +107,13 @@ func runQueryCmd(cmd *cobra.Command, args []string) {
 	// convert the query or sql file arg into an array of executable queries - check names queries in the current workspace
 	queries := execute.GetQueries(args, workspace)
 
+	clientReadyChannel := make(chan bool)
+
+	// go func() {
```
Comment (kaidaguerre) [inline cmd/query.go:112]: why not in prerun?

File: cmd/query.go [hunk 2]
```diff
@@ -122,14 +122,22 @@ func runQueryCmd(cmd *cobra.Command, args []string) {
 	err = db.CreateMetadataTables(workspace.GetResourceMaps(), client)
 	utils.FailOnError(err)
 
+	close(clientReadyChannel)
```
Comment (kaidaguerre) [inline cmd/query.go:125]: do we still need client ready channel?

File: cmd/service.go [hunk 3]
```diff
@@ -231,6 +234,27 @@ func runServiceStatusCmd(cmd *cobra.Command, args []string) {
...
+		defer func() {
+			close(gotDetails)
+			display.StopSpinner(sp)
+			printAllStatus(processes)
+		}()
+
+		processes, err = db.FindAllSteampipePostgresInstances()
+		if err != nil {
+			utils.ShowError(err)
+			return
```
Comment (kaidaguerre) [inline cmd/service.go:252]: won't the defer show an empty list of processes

File: cmd/service.go [hunk 4]
```diff
@@ -231,6 +234,27 @@ func runServiceStatusCmd(cmd *cobra.Command, args []string) {
...
+	} else if viper.GetBool(constants.ArgAll) {
+		var processes []*psutils.Process
+		var err error
+
+		gotDetails := make(chan bool)
+		sp := display.StartSpinnerAfterDelay("Getting details", constants.SpinnerShowTimeout, gotDetails)
+		defer func() {
+			close(gotDetails)
+			display.StopSpinner(sp)
+			printAllStatus(processes)
```
Comment (kaidaguerre) [inline cmd/service.go:246]: this code feels a little unclear, do we really need to put `printAllStatus` in the defer?

File: cmdconfig/builder.go [hunk 5]
```diff
@@ -18,6 +21,7 @@ func OnCmd(cmd *cobra.Command) *CmdBuilder {
 	cfg.bindings = map[string]*pflag.Flag{}
 
 	originalPreRun := cfg.cmd.PreRun
+	originalRun := cfg.cmd.Run
```
Comment (kaidaguerre) [inline cmdconfig/builder.go:24]: add comment to explain why we need to do this

File: db/client.go [hunk 6]
```diff
@@ -178,6 +189,8 @@ func (c *Client) ConnectionMap() *steampipeconfig.ConnectionMap {
 
 // return both the raw query result and a sanitised version in list form
 func (c *Client) loadSchema() {
+	utils.LogTime("db.loadSchema start")
+	defer utils.LogTime("db.loadSchema end")
```
Comment (kaidaguerre) [inline db/client.go:193]: lets always put a blank line after time logging statements

File: db/query.go [hunk 7]
```diff
@@ -23,9 +21,9 @@ func EnsureDbAndStartService(invoker Invoker) error {
 		return errors.New("could not retrieve service status")
 	}
 
-	if status != nil && status.Invoker != InvokerService {
-		return fmt.Errorf("You already have a %s session open. To run multiple sessions, first run %s.\nTo kill existing sessions run %s", constants.Bold(fmt.Sprintf("steampipe %s", status.Invoker)), constants.Bold("steampipe service start"), constants.Bold("steampipe service stop --force"))
-	}
+	// if status != nil && status.Invoker != InvokerService {
```
Comment (kaidaguerre) [inline db/query.go:24]: delete if not needed

File: db/stop_database.go [hunk 8]
```diff
@@ -27,20 +26,39 @@ const (
 	ServiceStopFailed
 	// ServiceStopTimedOut :: StopStatus - service stop attempt timed out
 	ServiceStopTimedOut
+	// ServiceStopFailed_IsImplicit :: StopStatus - when service is running implicit to a different command
```
Comment (kaidaguerre) [inline db/stop_database.go:29]: where are these used?

File: db/stop_database.go [hunk 9]
```diff
@@ -72,42 +104,30 @@ func StopDB(force bool, invoker Invoker) (StopStatus, error) {
...
-			if previousProcess != nil {
-				// we have an errant process
-				killProcessTree(previousProcess)
-				continue
+		defer func() {
+			close(checkedPreviousInstances)
+			display.StopSpinner(s)
+		}()
+		if processes, err := FindAllSteampipePostgresInstances(); err == nil {
+			for _, process := range processes {
```
Comment (kaidaguerre) [inline db/stop_database.go:112]: put in function killAllProcesses

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: cmd/query.go (hunk 2 above)
Comment (binaek89) [reply cmd/query.go]: Fixed

File: cmd/query.go (hunk 1 above)
Comment (binaek89) [reply cmd/query.go]: this is not required anymore. removed and fixed

File: cmd/service.go (hunk 3 above)
Comment (binaek89) [reply cmd/service.go:287]: fixed

File: cmd/service.go (hunk 4 above)
Comment (binaek89) [reply cmd/service.go:281]: refactored and fixed

File: cmdconfig/builder.go (hunk 5 above)
Comment (binaek89) [reply cmdconfig/builder.go:25]: comments added

File: db/client.go (hunk 6 above)
Comment (binaek89) [reply db/client.go:193]: 👍

File: db/query.go (hunk 7 above)
Comment (binaek89) [reply db/query.go]: 👍

File: db/stop_database.go (hunk 8 above)
Comment (binaek89) [reply db/stop_database.go]: This got left over from an earlier iteration. Removed

File: db/stop_database.go (hunk 9 above)
Comment (binaek89) [reply db/stop_database.go]: fixed

File: cmd/service.go [hunk 10]
```diff
@@ -143,20 +146,55 @@ func runServiceStartCmd(cmd *cobra.Command, args []string) {
...
+		if listen != info.ListenType {
+			utils.ShowError(fmt.Errorf("cannot convert service - listen binding does not match with running instance"))
+			return
+		}
+
+		// convert
+		info.Invoker = db.InvokerService
+		err = info.Save()
+		if err != nil {
+			utils.ShowErrorWithMessage(err, "conversion failed")
```
Comment (kaidaguerre) [inline cmd/service.go:175]: can we have a more verbose/clear error message - as a user, I have no idea what 'conversion' is

File: cmd/service.go [hunk 11]
```diff
@@ -143,20 +146,55 @@ func runServiceStartCmd(cmd *cobra.Command, args []string) {
...
+			printStatus(info)
+			return
+		}
+		// check that we have the same port and listen parameters
+		if port != info.Port {
+			utils.ShowError(fmt.Errorf("cannot convert service - port does not match with running instance"))
+			return
+		}
+		if listen != info.ListenType {
+			utils.ShowError(fmt.Errorf("cannot convert service - listen binding does not match with running instance"))
```
Comment (kaidaguerre) [inline cmd/service.go:167]: can we display the current and requested binding

File: cmd/service.go [hunk 12]
```diff
@@ -240,7 +282,60 @@ func runServiceStatusCmd(cmd *cobra.Command, args []string) {
 			fmt.Println("Steampipe database service is NOT running")
 		}
 	}
+}
+
+func showAllStatus() {
+	var processes []*psutils.Process
+	var err error
+
+	gotDetails := make(chan bool)
```
Comment (kaidaguerre) [inline cmd/service.go:291]: rename detailsChan

File: cmd/service.go [hunk 13]
```diff
@@ -240,7 +282,60 @@ func runServiceStatusCmd(cmd *cobra.Command, args []string) {
...
+	if err != nil {
+		utils.ShowError(err)
+		return
+	}
+
+	if len(processes) == 0 {
+		fmt.Println("There are no steampipe services running")
+		return
+	}
+	headers := []string{"#PID", "Install Directory", "Port", "Listen"}
```
Comment (kaidaguerre) [inline cmd/service.go:307]: remove `#`

File: cmd/service.go [hunk 14]
```diff
@@ -240,7 +282,60 @@ func runServiceStatusCmd(cmd *cobra.Command, args []string) {
...
+		rows = append(rows, []string{pid, installDir, port, string(listen)})
+	}
+
+	display.ShowWrappedTable(headers, rows, false)
+}
+
+func getServiceProcessDetails(process *psutils.Process) (string, string, string, db.StartListenType) {
+	cmdLine, _ := process.CmdlineSlice()
+
+	installDir := strings.TrimSuffix(cmdLine[0], "/db/12.1.0/postgres/bin/postgres")
```
Comment (kaidaguerre) [inline cmd/service.go:321]: should we use a const for this?

File: db/client_connections.go [hunk 15]
```diff
@@ -17,6 +17,8 @@ import (
 // and update the database schema and search path to reflect the required connections
 // return whether any changes have been made
 func (c *Client) RefreshConnections() (bool, error) {
+	utils.LogTime("db.RefreshConnections start")
+	defer utils.LogTime("db.RefreshConnections end")
```
Comment (kaidaguerre) [inline db/client_connections.go:21]: blank line

File: db/functions.go [hunk 16]
```diff
@@ -24,6 +25,8 @@ ORDER BY
 **/
 
 func refreshFunctions() error {
+	utils.LogTime("db.refreshFunctions start")
+	defer utils.LogTime("db.refreshFunctions end")
```
Comment (kaidaguerre) [inline db/functions.go:29]: blank line

File: cmd/query.go [hunk 17]
```diff
@@ -114,6 +107,10 @@ func runQueryCmd(cmd *cobra.Command, args []string) {
 	// convert the query or sql file arg into an array of executable queries - check names queries in the current workspace
 	queries := execute.GetQueries(args, workspace)
 
+	// start db if necessary
+	err = db.EnsureDbAndStartService(db.InvokerQuery)
```
Comment (kaidaguerre) [inline cmd/query.go:111]: why not in prerun as for other commands?

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: cmd/query.go (hunk 17 above)
Comment (binaek89) [reply cmd/query.go]: fixed.

commented previous code for now

File: db/functions.go (hunk 16 above)
Comment (binaek89) [reply db/functions.go:29]: fixed

File: db/client_connections.go (hunk 15 above)
Comment (binaek89) [reply db/client_connections.go:21]: fixed

File: cmd/service.go (hunk 13 above)
Comment (binaek89) [reply cmd/service.go]: Done

File: cmd/service.go (hunk 11 above)
Comment (binaek89) [reply cmd/service.go]: updated the messagings in the failures to be a little but more descriptive

File: cmd/service.go (hunk 10 above)
Comment (binaek89) [reply cmd/service.go]: updated see below

File: cmd/service.go (hunk 14 above)
Comment (binaek89) [reply cmd/service.go]: done

File: cmd/service.go [hunk 18]
```diff
@@ -301,22 +446,70 @@ func runServiceStopCmd(cmd *cobra.Command, args []string) {
...
+	var err error
+
 	force := cmdconfig.Viper().GetBool(constants.ArgForce)
-	status, err := db.StopDB(force, db.InvokerService)
+	if force {
+		status, err = db.StopDB(force, db.InvokerService)
+	} else {
+		info, err := db.GetStatus()
+		if err != nil {
+			utils.ShowError(fmt.Errorf("error during stop"))
```
[the remark this answers is not in the document]
Comment (kaidaguerre) [reply cmd/service.go:458]:
```
stop failed with error: xxxxxxx
```

File: db/pid_exists.go [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+// exists or not.
+func PidExists(pid int) (bool, error) {
+	// ps -p 27098 -o comm=
+	cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("ps -p %d -o comm=", pid))
+	o, err := cmd.Output()
+	if err != nil {
+		return false, nil
+	}
+	if strings.Contains(string(o), "(postgres)") {
+		// this means that postgres went away, but the process has remained.
```
Comment (kaidaguerre) [inline db/pid_exists.go:23]: let's update this comment slightly - this is open source after all

```
// this means that postgres went away, but the process has not yet completed.
// we are not sure why this occurs but can safely treat it as if the process does not exist
```

File: db/start_service.go [hunk 20]
```diff
@@ -14,9 +14,13 @@ import (
...
+// StartImplicitService :: invokes `steampipe service start --database-listen local --refresh=false --invoker query`
+func StartImplicitService(invoker Invoker) {
+	utils.LogTime("db.StartImplicitService start")
+	defer utils.LogTime("db.StartImplicitService end")
+
+	log.Println("[TRACE] start implicit service")
+
 	// spawn a process to start the service, passing refresh=false to ensure we DO NOT refresh connections
 	// (as we will do that ourselves)
 	cmd := exec.Command(os.Args[0], "service", "start", "--database-listen", "local", "--refresh=false", "--invoker", string(invoker), "--install-dir", constants.SteampipeDir)
```
Comment (kaidaguerre) [inline db/start_service.go:26]: still spawning?

File: db/stop_database.go [hunk 21]
```diff
@@ -32,15 +31,30 @@ const (
...
 	if client != nil {
 		client.Close()
 	}
 
 	status, _ := GetStatus()
 
-	// force stop if the service was invoked by the same invoker and we are the last one
-	if status != nil && status.Invoker == invoker {
+	if status != nil {
+		if status.Invoker == InvokerService {
```
Comment (kaidaguerre) [inline db/stop_database.go:44]: comment explaining logic here - why nothing to do?

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: cmd/service.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (binaek89) [reply cmd/service.go:458]: done

File: db/locations.go [hunk 22]
```diff
@@ -9,6 +9,8 @@ import (
 	"github.com/turbot/steampipe/utils"
 )
 
+const ServiceExecutableLocationRelativeToInstallDir = "db/12.1.0/postgres/bin/postgres"
```
[the remark this answers is not in the document]
Comment (binaek89) [reply db/locations.go:12]: renamed

File: db/locations.go [hunk 23]
```diff
@@ -76,3 +78,5 @@ func getPgHbaConfLocation() string {
 func getPasswordFileLocation() string {
 	return filepath.Join(getDatabaseLocation(), ".passwd")
 }
+
+const PostgresBinaryExecutableLocationRelativeToInstallDir = "/db/12.1.0/postgres/bin/postgres"
```
[the remark this answers is not in the document]
Comment (binaek89) [reply db/locations.go:82]: renamed

File: db/pid_exists.go (hunk 19 above)
Comment (binaek89) [reply db/pid_exists.go:23]: updated comment

File: db/start_service.go (hunk 20 above)
Comment (binaek89) [reply db/start_service.go]: pushed this

File: db/stop_database.go (hunk 21 above)
Comment (binaek89) [reply db/stop_database.go:44]: explained

File: cmd/service.go [hunk 24]
```diff
@@ -301,22 +446,70 @@ func runServiceStopCmd(cmd *cobra.Command, args []string) {
...
+			return
+		}
+		if info == nil {
+			fmt.Println("Service is not running")
+			return
+		}
+		if info.Invoker != db.InvokerService {
+			fmt.Printf(`
```
[the remark this answers is not in the document]
Comment (binaek89) [reply cmd/service.go]: done

**Pull request merged**